src/Formik.tsx

```diff
@@ -259,92 +252,94 @@ export function useFormik<Values = object>({
       props.validationSchema,
       runFieldLevelValidations,
       runValidateHandler,
       runValidationSchema,
-      state.errors,
-      state.values,
-      fields,
     ]
   );
 
-  React.useEffect(() => {
-    if (
-      prevState.values !== state.values &&
-      !!validateOnChange &&
-      !state.isSubmitting &&
-      isMounted.current != null
-    ) {
-      const [validate, cancel] = makeCancelable(validateForm());
-      validate.then(x => x).catch(x => x); // catch the rejection silently
-      return cancel;
-    }
-    return;
-  }, [
-    prevState.values,
-    state.isSubmitting,
-    state.values,
-    validateForm,
-    validateOnChange,
-    isMounted,
-  ]);
+  // Run validations and dispatching the result as low-priority via rAF.
+  //
+  // The thinking is that validation as a result of onChange and onBlur
+  // should never block user input. Note: This method should never be called
+  // during the submission phase because validation prior to submission
+  // is actaully high-priority since we absolutely need to guarantee the
+  // form is valid before executing props.onSubmit.
+  const validateFormWithLowPriority = React.useCallback(
+    (values: Values = state.values) => {
+      if (!hasValidationFns) {
+        // No validation functions, so we bail out immediately
+        return Promise.resolve(emptyErrors);
+      }
 
-  React.useEffect(() => {
-    if (
-      prevState.touched !== state.touched &&
-      !!validateOnBlur &&
-      !state.isSubmitting &&
-      isMounted.current != null
-    ) {
-      const [validate, cancel] = makeCancelable(validateForm());
-      validate.then(x => x).catch(x => x); // catch the rejection silently
-      return cancel;
-    }
-    return;
-  }, [
-    prevState.touched,
-    state.isSubmitting,
-    state.touched,
-    validateForm,
-    validateOnBlur,
-    isMounted,
-  ]);
+      return Promise.resolve().then(() => {
+        // TODO: replace rAF with react deferred update API when it's ready https://github.com/facebook/react/issues/13306
+        requestAnimationFrame(() => {
+          return runAllValidations(values).then(combinedErrors => {
+            if (isMounted.current != null) {
+              dispatch({ type: 'SET_ERRORS', payload: combinedErrors });
+            }
+            return combinedErrors;
+          });
+        });
+      });
+    },
+    [hasValidationFns, runAllValidations, state.values]
+  );
+
+  // Run all validations methods and update state accordingly
+  const validateFormWithHighPriority = React.useCallback(
+    (values: Values = state.values) => {
+      if (!hasValidationFns) {
+        // No validation functions, so we bail out immediately
+        return Promise.resolve(emptyErrors);
+      }
+      dispatch({ type: 'SET_ISVALIDATING', payload: true });
+      return runAllValidations(values).then(combinedErrors => {
+        if (
+          !isEqual(state.errors, combinedErrors) &&
+          isMounted.current != null
```

Comment:
if you switch order of this conditions, `isEqual` would not need to run if component is not mounted, thus slightly faster in edge cases. Correct?

Reply:
Correct.

Reply:
Do we even need this isEqual? I have lost track at a this point.

Reply:
Good point, probably not. I think we are good without it